### ISSUE TITLE
Pico Cluster Ansatz mit Chatgpt

### ISSUE THREAD

**Issue by Breleasy** [bmacharia/pico-cluster#1]
Mein neues Projekt

**Comment by Breleasy** [comment]
https://chatgpt.com/s/t_68f59d5bde148191b85b3ef3614b949a


> Mein neues Projekt
> 
> 
> 
>

**Comment by Breleasy** [comment]
Stark – wir ziehen’s jetzt durch: Live-Dashboard + RKNN-Inference-Demo, komplett „drag & drop“. Keine Spielerei, sondern praxistauglich.

Ich gebe dir neue/aktualisierte Files. Lege sie in deinen bestehenden rvcluster/-Ordner. Danach: pip install -r requirements.txt, python3 master_dashboard.py starten – du siehst Live-Status, Jobs, π-Demo, und kannst RKNN-Inference triggern (sofern rknn_runtime auf den Nodes vorhanden ist; sonst bekommst du eine klare Fehlermeldung statt Schweigen).

1) requirements.txt (ergänzt)

pyzmq msgpack numpy # Dashboard fastapi uvicorn[standard] # (Host: Training/Export – optional) torch onnx scikit-learn # (Host: Conversion – wenn installiert) # rknn-toolkit2 

2) Worker erweitern: worker.py (ersetzen)

Neu: echter npu_infer-Hook. Wenn rknn_runtime fehlt, wird sauber ein Fehler zurückgegeben.
MASTER_IP ersetzt weiterhin dein deploy.sh.

#!/usr/bin/env python3 import os, time, random, socket, importlib import zmq, msgpack MASTER_OUT = "tcp://MASTER_IP:5556" # Worker -> Master MASTER_IN = "tcp://MASTER_IP:5555" # Master -> Worker NODE = socket.gethostname() ctx = zmq.Context.instance() pull = ctx.socket(zmq.PULL); pull.connect(MASTER_IN) push = ctx.socket(zmq.PUSH); push.connect(MASTER_OUT) def send(obj): obj = dict(obj); obj["node"] = NODE push.send(msgpack.packb(obj, use_bin_type=True)) # ---- CPU-Demos ---- def simulate_pi(n:int): inside = 0 for _ in range(int(n)): x, y = random.random(), random.random() inside += (x*x + y*y) <= 1.0 return {"inside": inside, "total": int(n)} def moving_avg(series, w:int=5): out=[]; acc=0.0; buf=[] for v in series: v=float(v); buf.append(v); acc+=v if len(buf)>w: acc-=buf.pop(0) out.append(acc/len(buf)) return {"series": out} # ---- RKNN/NPU Inference ---- def npu_infer(model_path:str, input_path:str=None): try: rknn_runtime = importlib.import_module("rknn.api.rknn_runtime") except Exception as e: return {"error":"rknn_runtime_not_available", "detail":repr(e)} from rknn.api import RKNNRuntime import numpy as np # Dummy-Input, wenn kein Bild angegeben: if input_path and os.path.exists(input_path): # Minimal: Numpy-Laden erwartet .npy als Beispiel # (für echte Bild-Pfade bitte Pillow+preprocess nutzen) x = np.load(input_path) else: x = np.random.rand(1,224,224,3).astype("float32") try: rknn = RKNNRuntime() ret = rknn.load_model(model=model_path) if ret != 0: return {"error":"load_model_failed", "ret":int(ret)} ret = rknn.init_runtime() if ret != 0: return {"error":"init_runtime_failed", "ret":int(ret)} outs = rknn.inference(inputs=[x]) rknn.release() # Kompakt zurückgeben (nur Statistik, nicht volle Tensoren) return {"ok": True, "out_shapes":[list(o.shape) for o in outs], "out_sample": float(outs[0].ravel()[0])} except Exception as e: return {"error":"npu_exception", "detail":repr(e)} def handle(job:dict): t = job.get("type") if t == "simulate_pi": return simulate_pi(job.get("n", 1000000)) if t == "preprocess_ma": return moving_avg(job.get("series", []), int(job.get("w",5))) if t == "npu_infer": return npu_infer(job.get("model","/root/model.rknn"), job.get("input")) if t == "ping": return {"pong": True} return {"error":"unknown_job"} # --- Loop --- send({"type":"hello", "cap":["simulate_pi","preprocess_ma","npu_infer"]}) last_hb = 0 while True: # Heartbeat zum Master alle 5 s now = time.time() if now - last_hb > 5: send({"type":"hb", "ts": now}) last_hb = now try: job = msgpack.unpackb(pull.recv(flags=zmq.NOBLOCK), raw=False) except zmq.Again: time.sleep(0.02) continue tgt = job.get("target") if tgt and tgt != NODE: # simple targeting continue jid = job.get("id", 0) try: res = handle(job) send({"type":"result", "id": jid, "res": res}) except Exception as e: send({"type":"result", "id": jid, "res": {"error": repr(e)}}) 

3) Master + Dashboard in einem Prozess: master_dashboard.py (neu)

Startet ZeroMQ-Master und ein FastAPI-Dashboard.
WebSocket pusht Live-Events (Nodes online, Heartbeats, Ergebnisse).
HTTP-Endpoints zum Starten von π-Demo & NPU-Jobs.

#!/usr/bin/env python3 import time, threading, queue, json from pathlib import Path from typing import Dict, List, Tuple import zmq, msgpack from fastapi import FastAPI, WebSocket, WebSocketDisconnect from fastapi.responses import HTMLResponse import uvicorn MASTER_BIND_IN = "tcp://0.0.0.0:5556" # Worker -> Master MASTER_BIND_OUT = "tcp://0.0.0.0:5555" # Master -> Worker # ------------ Cluster Core ------------ _job_id = 0 def next_job_id(): global _job_id _job_id += 1 return _job_id class Cluster: def __init__(self, conf_path="cluster.conf"): self.nodes: List[Tuple[str,str]] = [] self.node_alive: Dict[str,float] = {} self.node_caps: Dict[str,List[str]] = {} self.results = queue.Queue() self.events = queue.Queue() # for WS broadcast self.ctx = zmq.Context.instance() self.push = self.ctx.socket(zmq.PUSH); self.push.bind(MASTER_BIND_OUT) self.pull = self.ctx.socket(zmq.PULL); self.pull.bind(MASTER_BIND_IN) self.load_conf(conf_path) def load_conf(self, path): with open(path) as f: for line in f: line=line.strip() if not line or line.startswith("#"): continue name, ip = line.split() self.nodes.append((name, ip)) def rx_loop(self): while True: msg = self.pull.recv() data = msgpack.unpackb(msg, raw=False) n = data.get("node","?") t = data.get("type","") ts = time.time() if t in ("hello","hb"): self.node_alive[n] = ts if t == "hello": self.node_caps[n] = data.get("cap", []) if t == "result": self.results.put(data) # push every event to WS queue self.events.put({"ts":ts, "event":t, "data":data}) def start_rx(self): t = threading.Thread(target=self.rx_loop, daemon=True) t.start() def broadcast(self, payload:dict): self.push.send(msgpack.packb(payload, use_bin_type=True)) def send_job(self, node_name, payload:dict): payload = dict(payload) payload["target"] = node_name self.broadcast(payload) def live_status(self): now = time.time() return { "nodes":[{"name":n, "ip":ip, "alive": now - self.node_alive.get(n,0) < 10, "last_seen": self.node_alive.get(n,0), "cap": self.node_caps.get(n,[])} for n,ip in self.nodes], "utc": now } # ------------ Dashboard (FastAPI) ------------ app = FastAPI() cluster = Cluster("cluster.conf") cluster.start_rx() cluster.broadcast({"id": next_job_id(), "type":"ping"}) @app.get("/") def home(): return HTMLResponse(Path(__file__).with_name("ui.html").read_text(encoding="utf-8")) @app.get("/api/status") def api_status(): return cluster.live_status() @app.post("/api/run/pi") def api_run_pi(total: int = 10_000_000): N = len(cluster.nodes) if N == 0: return {"error":"no_nodes"} per = total // N expected = set() for name,_ in cluster.nodes: jid = next_job_id() expected.add(jid) cluster.send_job(name, {"id": jid, "type": "simulate_pi", "n": per}) # Sammeln im Hintergrund-Thread, Ergebnis an WS streamen: def collector(exp): inside = total_s = 0 while exp: res = cluster.results.get() if res.get("type")!="result": continue jid = res.get("id") if jid in exp: exp.remove(jid) r = res.get("res", {}) inside += r.get("inside", 0) total_s += r.get("total", 0) cluster.events.put({"ts":time.time(),"event":"pi_progress","data":{"inside":inside,"total":total_s}}) pi = 4.0 * inside / max(1,total_s) cluster.events.put({"ts":time.time(),"event":"pi_done","data":{"pi":pi,"total":total_s}}) threading.Thread(target=collector, args=(expected,), daemon=True).start() return {"ok": True, "message": f"π-Run started with total={total}"} @app.post("/api/run/npu") def api_run_npu(model: str="/root/model.rknn", node: str="node01", input_path: str=None): jid = next_job_id() cluster.send_job(node, {"id": jid, "type": "npu_infer", "model": model, "input": input_path}) def wait_one(jid_): while True: res = cluster.results.get() if res.get("type")=="result" and res.get("id")==jid_: cluster.events.put({"ts":time.time(),"event":"npu_result","data":res}) break threading.Thread(target=wait_one, args=(jid,), daemon=True).start() return {"ok": True, "message": f"npu job sent to {node}", "job_id": jid} # WebSocket für Live-Events clients: List[WebSocket] = [] @app.websocket("/ws") async def ws(ws: WebSocket): await ws.accept() clients.append(ws) # Beim Connect sofort Status pushen await ws.send_text(json.dumps({"type":"status","data":cluster.live_status()})) try: while True: # events non-blocking aus queue lesen try: evt = cluster.events.get(timeout=0.2) # broadcast an alle dead=[] for c in clients: try: await c.send_text(json.dumps({"type":evt.get("event"), "data":evt.get("data")})) except: dead.append(c) for d in dead: if d in clients: clients.remove(d) except queue.Empty: pass # Client-Eingaben (optional, ignorieren) try: await ws.receive_text() except: pass except WebSocketDisconnect: if ws in clients: clients.remove(ws) if __name__ == "__main__": print("Master+Dashboard läuft auf http://0.0.0.0:8000") uvicorn.run(app, host="0.0.0.0", port=8000) 

4) Mini-UI: ui.html (neu)

Minimalistisch, aber klar: Node-Status, Buttons für π-Run und NPU-Job, Live-Events via WebSocket.

<!doctype html> <html> <head> <meta charset="utf-8"> <title>RV1106 Cluster Dashboard</title> <style> body { font-family: system-ui, sans-serif; margin: 24px; } .row { display:flex; gap:24px; align-items:flex-start; } .card { border:1px solid #ddd; border-radius:12px; padding:16px; } .nodes table { border-collapse: collapse; } .nodes th, .nodes td { border:1px solid #ddd; padding:6px 8px; } .ok { color: #107010; font-weight:600; } .bad { color: #a01010; font-weight:600; } button { padding:8px 14px; border-radius:8px; border:1px solid #333; cursor:pointer;} pre { background:#f7f7f7; padding:12px; border-radius:8px; max-height:220px; overflow:auto;} </style> </head> <body> <h1>RV1106 Cluster Dashboard</h1> <div class="row"> <div class="card nodes"> <h3>Nodes</h3> <table id="nodes"><thead><tr><th>Name</th><th>IP</th><th>Alive</th><th>Caps</th><th>Last seen</th></tr></thead><tbody></tbody></table> </div> <div class="card"> <h3>Aktionen</h3> <button onclick="runPi()">Parallel-π (10 Mio)</button> <div style="height:8px"></div> <input id="npuNode" value="node01" /> <input id="model" value="/root/model.rknn" /> <button onclick="runNPU()">NPU-Infer</button> <div style="height:8px"></div> <small>Hinweis: Für NPU muss <code>rknn_runtime</code> auf dem Node installiert sein.</small> </div> </div> <div class="card" style="margin-top:16px"> <h3>Events</h3> <pre id="log"></pre> </div> <script> const logEl = document.getElementById('log'); const nodesBody = document.querySelector('#nodes tbody'); function log(msg){ logEl.textContent += msg + "\n"; logEl.scrollTop = logEl.scrollHeight; } async function refreshStatus(){ const r = await fetch('/api/status'); const j = await r.json(); nodesBody.innerHTML = ''; for (const n of j.nodes){ const tr = document.createElement('tr'); tr.innerHTML = `<td>${n.name}</td><td>${n.ip}</td> <td>${n.alive?'<span class="ok">alive</span>':'<span class="bad">down</span>'}</td> <td>${(n.cap||[]).join(', ')}</td> <td>${new Date(n.last_seen*1000).toLocaleTimeString()}</td>`; nodesBody.appendChild(tr); } } refreshStatus(); const ws = new WebSocket(`ws://${location.host}/ws`); ws.onmessage = (ev)=>{ const m = JSON.parse(ev.data); if (m.type==='status'){ refreshStatus(); return; } if (m.type==='pi_progress'){ log(`π progress: ${m.data.inside}/${m.data.total}`); return; } if (m.type==='pi_done'){ log(`π done: π ≈ ${m.data.pi.toFixed(6)} (samples=${m.data.total})`); refreshStatus(); return; } if (m.type==='npu_result'){ log(`NPU result from ${m.data.node}: ${JSON.stringify(m.data.res).slice(0,200)}...`); return; } if (m.type==='hb'){ /* ignore spam */ return; } if (m.type==='hello'){ log(`HELLO from ${m.data.node} caps=[${m.data.cap}]`); refreshStatus(); return; } if (m.type==='result'){ log(`Result from ${m.data.node}: ${JSON.stringify(m.data.res).slice(0,200)}...`); return; } log(`Event ${m.type}: ${JSON.stringify(m.data).slice(0,200)}...`); }; async function runPi(){ await fetch('/api/run/pi', {method:'POST'}); log('π-Run gestartet…'); } async function runNPU(){ const node = document.getElementById('npuNode').value; const model = document.getElementById('model').value; await fetch(`/api/run/npu?node=${encodeURIComponent(node)}&model=${encodeURIComponent(model)}`, {method:'POST'}); log(`NPU-Job an ${node} gesendet…`); } </script> </body> </html> 

5) RKNN-Demo am Host (optional): rknn_convert_demo.py (neu)

Ehrlich: Konvertierung läuft am X270. Auf dem Node brauchst du nur model.rknn + rknn_runtime.

#!/usr/bin/env python3 # Beispiel-Flow: ONNX -> RKNN (nur wenn rknn-toolkit2 installiert ist) import numpy as np import os, sys try: from rknn.api import RKNN except Exception as e: print("rknn-toolkit2 nicht installiert:", e) sys.exit(1) ONNX_MODEL = "mobilenet_v1.onnx" # Beispiel RKNN_MODEL = "model.rknn" def main(): rknn = RKNN() # Optional: Quantisierung etc. je nach Bedarf tunen rknn.config(mean_values=[[127.5,127.5,127.5]], std_values=[[127.5,127.5,127.5]]) print("Lade ONNX…") ret = rknn.load_onnx(model=ONNX_MODEL) if ret != 0: raise SystemExit("Load ONNX failed") print("Baue Graph…") ret = rknn.build(do_quantization=False) if ret != 0: raise SystemExit("Build failed") print("Speichere RKNN…") ret = rknn.export_rknn(RKNN_MODEL) if ret != 0: raise SystemExit("Export failed") print("OK ->", RKNN_MODEL) if __name__ == "__main__": main() 

Deploy auf Node (z. B. node01):

scp model.rknn root@192.168.77.101:/root/ # (Optional) Dummy-Input als .npy: python3 - <<'PY' import numpy as np; np.save("dummy.npy", np.random.rand(1,224,224,3).astype("float32")) PY scp dummy.npy root@192.168.77.101:/root/ 

Dashboard → NPU-Infer klicken → Ergebnis erscheint im Event-Log.
Falls rknn_runtime fehlt, siehst du eine klare Fehlermeldung – kein Black-Hole.

6) Start – in 3 Befehlen

# 1) Abhängigkeiten pip3 install -r requirements.txt # 2) Worker verteilen (setzt MASTER_IP in worker.py) bash deploy.sh 192.168.77.1 # 3) Dashboard + Master python3 master_dashboard.py # -> Öffne http://<X270-IP>:8000 

Klartext zur Realität

Training (schwer) bleibt auf dem X270.

Inference (leicht bis mittel) kann parallel auf den 10 RV1106-Nodes laufen; mit RKNN schöpfst du die NPU aus.

ZeroMQ + FastAPI ist schlank und stabil; kein Overkill wie Ray oder Kubernetes.

Wenn ein Node zickt, siehst du’s sofort im Dashboard (Heartbeat/Last-Seen), kein Rätselraten.

Wenn du möchtest, lege ich dir als nächstes noch einen einfachen Front-Panel-Button-Handler (z. B. „Hard Stop“, „All Ping“, „All Reboot“) sowie Job-Queues mit Prioritäten dazu.
